nexus/src/sagas.rs

```diff
@@ -184,72 +184,188 @@ async fn sic_alloc_server(
+                "default interface for {}",
+                saga_params.create_params.identity.name,
+            ),
+        },
+        vpc_name: default_name.clone(),
+        subnet_name: default_name.clone(),
+        ip: None, // Request an IP address allocation
+    };
     let vpc = osagactx
         .datastore()
-        .vpc_fetch_by_name(&params.project_id, &default_name)
+        .vpc_fetch_by_name(
+            &saga_params.project_id,
+            &internal_default_name.clone(),
+        )
         .await
         .map_err(ActionError::action_failed)?;
     let subnet = osagactx
         .datastore()
-        .vpc_subnet_fetch_by_name(&vpc.id(), &default_name)
+        .vpc_subnet_fetch_by_name(&vpc.id(), &internal_default_name)
         .await
         .map_err(ActionError::action_failed)?;
 
     let mac = db::model::MacAddr::new().map_err(ActionError::action_failed)?;
     let interface_id = Uuid::new_v4();
-    // Request an allocation
-    let ip = None;
     let interface = db::model::IncompleteNetworkInterface::new(
         interface_id,
         instance_id,
-        // TODO-correctness: vpc_id here is used for name uniqueness. Should
-        // interface names be unique to the subnet's VPC or to the
-        // VPC associated with the instance's default interface?
         vpc.id(),
-        subnet,
+        subnet.clone(),
         mac,
-        ip,
-        params::NetworkInterfaceCreate {
-            identity: IdentityMetadataCreateParams {
-                // By naming the interface after the instance id, we should
-                // avoid name conflicts on creation.
-                name: format!("default-{}", instance_id).parse().unwrap(),
-                description: format!(
-                    "default interface for {}",
-                    params.create_params.identity.name
-                ),
-            },
-        },
-    );
-
+        interface_params.identity.clone(),
+        interface_params.ip,
+    )
+    .map_err(ActionError::action_failed)?;
     let interface = osagactx
         .datastore()
         .instance_create_network_interface(interface)
         .await
+        .map_err(db::subnet_allocation::NetworkInterfaceError::into_external)
         .map_err(ActionError::action_failed)?;
-    Ok(interface.into())
+    Ok(Some(vec![interface.into()]))
 }
 
-async fn sic_create_network_interface_undo(
+async fn sic_create_network_interfaces_undo(
     sagactx: ActionContext<SagaInstanceCreate>,
 ) -> Result<(), anyhow::Error> {
     let osagactx = sagactx.user_data();
-    let network_interface =
-        sagactx.lookup::<NetworkInterface>("network_interface")?;
-
-    osagactx
-        .datastore()
-        .instance_delete_network_interface(&network_interface.identity.id)
-        .await?;
+    let network_interfaces = sagactx
+        .lookup::<Option<Vec<NetworkInterface>>>("network_interfaces")?;
+
+    // No networking at all created
+    if network_interfaces.is_none() {
```

Comment:
That's a good point, thanks for catching. I definitely missed the case where we fail partway through allocating multiple interfaces. I'll update it now.

Reply:
Ok, so I believe I've abused sagas by creating this Nexus API.

In this new API, one can create any number of NICs for an instance. Suppose the client specifies the _exact_ same NIC twice. This will obviously fail, since there's a name conflict, IP conflict, etc. However, the saga executor will only run the undo actions for those nodes that have successfully completed. (That's my understanding, @davepacheco correct me if I'm wrong.) That means that we'll have the first NIC still created, and no real way to delete it.

I'm going through the `steno` source to make sure I understand this. It seems like there _might_ be a way for the undo action to be taken when the forward action fails, but I'm not sure yet.

Reply:
So, one workaround I see is actually storing an inner `Result` as the output of the NIC creation saga node. `sic_create_network_interfaces` would return a `Result<Result<...>, ActionError>`, and the next step in the saga would check this and immediately fail. That way the actual `nic_create_network_interfaces_undo` function would run.

That feels _super_ gross, but it also feels pretty constraining to limit clients to only create zero or one NIC for an instance at the time of provision.

Reply:
I just sketched this out. It's definitely gross, but it's not the worst thing. We get some `??` or other shenanigans to handle a `Result<Result<...>>`. It also induces more coupling between the nodes. The communication is still "through" the serialized saga node output, but the downstream node has to do something special based on the output. That's always true, but feels more brittle here.

Reply:
Offline, @davepacheco had a simpler solution. I'm planning to add a "dummy" saga node before the NIC create node, that pretty much does nothing. This will have the undo action of deleting all NICs associated with the instance. That way, even if the NIC-creation node fails halfway through a list of NICs, the undo action is always done.

This might be an abstraction that's useful to build into Steno at some point. Either an unconditional undo, or an undo that's run even if the node fails. Another idea Dave floated is a kind of sub-saga, in which we can indicate that the executor should run N "copies" of this node in parallel, each with the same input, and serializing the ouptut to a list of some kind. It could then run the corresponding undo action for the successfully-completed nodes.